File: packages/csharp/src/Models/TransactionOrder.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace Tesote.Sdk.Models;

/// <summary>Source-account stub embedded in a <see cref="TransactionOrder"/>.</summary>
public sealed record TransactionOrderSourceAccount(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("payment_method_id")] string PaymentMethodId);

/// <summary>Destination metadata embedded in a <see cref="TransactionOrder"/>.</summary>
public sealed record TransactionOrderDestination(
    [property: JsonPropertyName("payment_method_id")] string PaymentMethodId,
    [property: JsonPropertyName("counterparty_id")] string? CounterpartyId,
    [property: JsonPropertyName("counterparty_name")] string? CounterpartyName);

/// <summary>Optional fee charged on a <see cref="TransactionOrder"/>.</summary>
public sealed record TransactionOrderFee(
    [property: JsonPropertyName("amount")] decimal Amount,
    [property: JsonPropertyName("currency")] string Currency);

/// <summary>Underlying bank transaction created from a successful order.</summary>
public sealed record TransactionOrderTesoteTransaction(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("status")] string Status);

/// <summary>Most recent submission attempt for a <see cref="TransactionOrder"/>.</summary>
public sealed record TransactionOrderLatestAttempt(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("attempt_number")] int AttemptNumber,
    [property: JsonPropertyName("external_reference")] string? ExternalReference,
    [property: JsonPropertyName("submitted_at")] string? SubmittedAt,
    [property: JsonPropertyName("completed_at")] string? CompletedAt,
    [property: JsonPropertyName("error_code")] string? ErrorCode,
    [property: JsonPropertyName("error_message")] string? ErrorMessage);

/// <summary>Beneficiary payload accepted by POST /v2/.../transaction_orders when no payment-method id is supplied.</summary>
public sealed record Beneficiary(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("bank_code")] string? BankCode,
    [property: JsonPropertyName("account_number")] string? AccountNumber,
    [property: JsonPropertyName("identification_type")] string? IdentificationType,
    [property: JsonPropertyName("identification_number")] string? IdentificationNumber);

/// <summary>Single transfer order — the unit of payment in v2.</summary>
public sealed record TransactionOrder(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("amount")] decimal Amount,
    [property: JsonPropertyName("currency")] string Currency,
    [property: JsonPropertyName("description")] string Description,
    [property: JsonPropertyName("reference")] string? Reference,
    [property: JsonPropertyName("external_reference")] string? ExternalReference,
    [property: JsonPropertyName("idempotency_key")] string? IdempotencyKey,
    [property: JsonPropertyName("batch_id")] string? BatchId,
    [property: JsonPropertyName("scheduled_for")] string? ScheduledFor,
    [property: JsonPropertyName("approved_at")] string? ApprovedAt,
    [property: JsonPropertyName("submitted_at")] string? SubmittedAt,
    [property: JsonPropertyName("completed_at")] string? CompletedAt,
    [property: JsonPropertyName("failed_at")] string? FailedAt,
    [property: JsonPropertyName("cancelled_at")] string? CancelledAt,
    [property: JsonPropertyName("source_account")] TransactionOrderSourceAccount SourceAccount,
    [property: JsonPropertyName("destination")] TransactionOrderDestination Destination,
    [property: JsonPropertyName("fee")] TransactionOrderFee? Fee,
    [property: JsonPropertyName("execution_strategy")] string? ExecutionStrategy,
    [property: JsonPropertyName("tesote_transaction")] TransactionOrderTesoteTransaction? TesoteTransaction,
    [property: JsonPropertyName("latest_attempt")] TransactionOrderLatestAttempt? LatestAttempt,
    [property: JsonPropertyName("metadata")] JsonObject? Metadata,
    [property: JsonPropertyName("created_at")] string CreatedAt,
    [property: JsonPropertyName("updated_at")] string UpdatedAt);

/// <summary>Response envelope for GET /v2/.../transaction_orders.</summary>
public sealed record TransactionOrderListResponse(
    [property: JsonPropertyName("items")] IReadOnlyList<TransactionOrder> Items,
    [property: JsonPropertyName("has_more")] bool HasMore,
    [property: JsonPropertyName("limit")] int Limit,
    [property: JsonPropertyName("offset")] int Offset);

/// <summary>Single error entry returned from POST /v2/.../batches.</summary>
public sealed record BatchOrderError(
    [property: JsonPropertyName("index")] int? Index,
    [property: JsonPropertyName("error")] string? Error,
    [property: JsonPropertyName("error_code")] string? ErrorCode);

/// <summary>Response envelope for POST /v2/.../batches.</summary>
public sealed record BatchCreateResponse(
    [property: JsonPropertyName("batch_id")] string BatchId,
    [property: JsonPropertyName("orders")] IReadOnlyList<TransactionOrder> Orders,
    [property: JsonPropertyName("errors")] IReadOnlyList<BatchOrderError> Errors);

/// <summary>Per-status order counts inside a batch.</summary>
public sealed record BatchStatusCounts(
    [property: JsonPropertyName("draft")] int Draft,
    [property: JsonPropertyName("pending_approval")] int PendingApproval,
    [property: JsonPropertyName("approved")] int Approved,
    [property: JsonPropertyName("processing")] int Processing,
    [property: JsonPropertyName("completed")] int Completed,
    [property: JsonPropertyName("failed")] int Failed,
    [property: JsonPropertyName("cancelled")] int Cancelled);

/// <summary>Response envelope for GET /v2/.../batches/{batch_id}.</summary>
public sealed record BatchSummary(
    [property: JsonPropertyName("batch_id")] string BatchId,
    [property: JsonPropertyName("total_orders")] int TotalOrders,
    [property: JsonPropertyName("total_amount_cents")] long TotalAmountCents,
    [property: JsonPropertyName("amount_currency")] string AmountCurrency,
    [property: JsonPropertyName("statuses")] BatchStatusCounts Statuses,
    [property: JsonPropertyName("batch_status")] string BatchStatus,
    [property: JsonPropertyName("created_at")] string CreatedAt,
    [property: JsonPropertyName("orders")] IReadOnlyList<TransactionOrder> Orders);

/// <summary>Response envelope for POST /v2/.../batches/{batch_id}/approve.</summary>
public sealed record BatchApproveResponse(
    [property: JsonPropertyName("approved")] int Approved,
    [property: JsonPropertyName("failed")] int Failed);

/// <summary>Response envelope for POST /v2/.../batches/{batch_id}/submit.</summary>
public sealed record BatchSubmitResponse(
    [property: JsonPropertyName("enqueued")] int Enqueued,
    [property: JsonPropertyName("failed")] int Failed);

/// <summary>Response envelope for POST /v2/.../batches/{batch_id}/cancel.</summary>
public sealed record BatchCancelResponse(
    [property: JsonPropertyName("cancelled")] int Cancelled,
    [property: JsonPropertyName("skipped")] int Skipped,
    [property: JsonPropertyName("errors")] IReadOnlyList<BatchOrderError> Errors);

File: packages/php/src/Models/TransactionOrder.php
<?php

declare(strict_types=1);

namespace Tesote\Sdk\Models;

/** TransactionOrder model (v2). */
final readonly class TransactionOrder
{
    public function __construct(
        public string $id,
        public string $status,
        public string $amount,
        public string $currency,
        public string $description,
        public ?string $reference,
        public ?string $externalReference,
        public ?string $idempotencyKey,
        public ?string $batchId,
        public ?string $scheduledFor,
        public ?string $approvedAt,
        public ?string $submittedAt,
        public ?string $completedAt,
        public ?string $failedAt,
        public ?string $cancelledAt,
        public SourceAccount $sourceAccount,
        public Destination $destination,
        public ?Money $fee,
        public ?string $executionStrategy,
        public ?TesoteTransactionRef $tesoteTransaction,
        public ?TransactionOrderAttempt $latestAttempt,
        public string $createdAt,
        public string $updatedAt,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $source = is_array($data['source_account'] ?? null) ? $data['source_account'] : [];
        $dest = is_array($data['destination'] ?? null) ? $data['destination'] : [];
        $fee = is_array($data['fee'] ?? null) ? $data['fee'] : null;
        $tesoteTxn = is_array($data['tesote_transaction'] ?? null) ? $data['tesote_transaction'] : null;
        $attempt = is_array($data['latest_attempt'] ?? null) ? $data['latest_attempt'] : null;

        return new self(
            id: (string) ($data['id'] ?? ''),
            status: (string) ($data['status'] ?? ''),
            amount: (string) ($data['amount'] ?? '0'),
            currency: (string) ($data['currency'] ?? ''),
            description: (string) ($data['description'] ?? ''),
            reference: isset($data['reference']) ? (string) $data['reference'] : null,
            externalReference: isset($data['external_reference']) ? (string) $data['external_reference'] : null,
            idempotencyKey: isset($data['idempotency_key']) ? (string) $data['idempotency_key'] : null,
            batchId: isset($data['batch_id']) ? (string) $data['batch_id'] : null,
            scheduledFor: isset($data['scheduled_for']) ? (string) $data['scheduled_for'] : null,
            approvedAt: isset($data['approved_at']) ? (string) $data['approved_at'] : null,
            submittedAt: isset($data['submitted_at']) ? (string) $data['submitted_at'] : null,
            completedAt: isset($data['completed_at']) ? (string) $data['completed_at'] : null,
            failedAt: isset($data['failed_at']) ? (string) $data['failed_at'] : null,
            cancelledAt: isset($data['cancelled_at']) ? (string) $data['cancelled_at'] : null,
            sourceAccount: SourceAccount::fromArray($source),
            destination: Destination::fromArray($dest),
            fee: $fee !== null ? Money::fromArray($fee) : null,
            executionStrategy: isset($data['execution_strategy']) ? (string) $data['execution_strategy'] : null,
            tesoteTransaction: $tesoteTxn !== null ? TesoteTransactionRef::fromArray($tesoteTxn) : null,
            latestAttempt: $attempt !== null ? TransactionOrderAttempt::fromArray($attempt) : null,
            createdAt: (string) ($data['created_at'] ?? ''),
            updatedAt: (string) ($data['updated_at'] ?? ''),
        );
    }
}
